### PR TITLE
Allow additional JVM args to be passed to jdt.ls

### PR DESCRIPTION
**Pull request by mjaewon**
This PR solves https://github.com/ycm-core/YouCompleteMe/issues/3288.

# Problem

Currently there is no way to pass additional JVM arguments to `jdt.ls`.

This means that `jdt.ls` doesn't understand java codes using [Lombok](https://projectlombok.org/), which needs to be passed to jvm as a `-javaagent`.

When `.ycm_extra_conf.py` is not given, `jdt.ls` doesn't understand lombok like the attached screenshot below:
before-fix

# Tests

Passes the existing tests:
```
$ LANG=en_GB.utf8 ./run_tests.py
...
=== 1837 passed, 50 skipped, 15 warnings, 1 rerun in 501.02s (0:08:21) ===
```

Integration test is done with my fork of https://github.com/mjaewon/YouCompleteMe + vim:

https://user-images.githubusercontent.com/1889939/106422461-767d2000-6413-11eb-84f6-39616eb5f022.mov

# Why this PR is helpful to others too

1. Lombok can be used.
2. It would be possible to optimize jdt.ls for example like `-noverify -Xmx1G -XX:+UseG1GC -XX:+UseStringDeduplication`, just like [its sample](https://www.eclipse.org/community/eclipse_newsletter/2017/may/article4.php).

---
This change is [Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)

### CODE REVIEW

**Comment by bstaletic** [comment]
Hi. Thanks for the pull request. As far as we're aware, additional jvm options can be passed using `JAVA_TOOL_OPTIONS` environment variable. Does that not work for you?

**Comment by mjaewon** [comment]
Hi, thanks for checking it quickly.

Unfortunately it seems that another JVM-based application like Gradle, which is automatically launched by `jdt.ls`, gets messed up when `JAVA_TOOL_OPTIONS` is defined in `~/.vimrc` like below:

```vim
let $JAVA_TOOL_OPTIONS = '-javaagent:{path_to_lombok}'
```

So we would need a workaround to specify jvm args separately for jdtls and gradle.

**Review by puremourning — COMMENTED** [review]
Yeah I think it’s ok to expose an option for this. A number of users have requested it. But we should not include the gradle wrapper in this repo, it’s just spam that every user needs to download. We should Just include the build.gradle if we need gradle

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MSReNeDEzB4jIaDoIrU:b-k6h9o1)* status: 0 of 2 LGTMs obtained

**Review by bstaletic — CHANGES_REQUESTED** [review]
> Reviewed 1 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @mjaewon)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSRg0lt5SDljQKGOkTr:b-fn96tg) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
> ```Python
>     if module:
>       settings = self._GetSettings( module, request_data[ 'extra_conf_data' ] )
>       return ( settings.get( 'jvm' ) or {} ).get( 'args' ) or []
> ```

Any reason for this nested lookup? Is there anything else we could expect to put in `'jvm'`'s value in the future?
___
*[ycmd/completers/java/java_completer.py, line 441 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRfoUt1WkLxZSFq4Zy:-MSRfoUt1WkLxZSFq4Zz:b-fifxv9) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L441)):*
> ```Python
> 
> 
>   def _GetSettings( self, module, client_data ):
> ```

There's already a `GetSettings()` in the base class. You don't need to reinvent it.

**Review by puremourning — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @mjaewon)

___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MSRipff3zgWtlTbkRqt:bpa0d6n) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
> ```Python
> 
> PATH_TO_LOMBOK = os.path.expanduser(
>   '~/.gradle/caches/modules-2/files-2.1/org.projectlombok/lombok/' +
> ```

This seems hacky. Feels like it’s easy to break. Can we have CI dowload the jar to a known location, or something?
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 13 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRj45r0EGZZEorsQV-:-MSRj45r0EGZZEorsQV0:buhz2dn) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L13)):*
> ```Python
> 
>   jvm_args = [
>     '-noverify',
> ```

What’s noverify do?

**Comment by mjaewon** [comment]
> Yeah I think it’s ok to expose an option for this. A number of users have requested it. But we should not include the gradle wrapper in this repo, it’s just spam that every user needs to download. We should Just include the build.gradle if we need gradle

@puremourning Thanks for pointing this out. Removed gradle wrappers in this commit a142e46.

I've made a separate commit for that, but let me know if you'd prefer rebase/force-push. First time here, so not sure what's preferred here.

**Comment by mjaewon** [comment]
@bstaletic Thanks for your review.

> Any reason for this nested lookup? Is there anything else we could expect to put in `'jvm'`'s value in the future?

Not really. I will make it flat like `jvm_args` if you'd think that is better.

> There's already a `GetSettings()` in the base class. You don't need to reinvent it.

Thanks for pointing this out. Fixed in 758dd6d.

**Comment by mjaewon** [comment]
> This seems hacky. Feels like it’s easy to break. Can we have CI dowload the jar to a known location, or something?

It's technically possible, but I'm afraid that it would make code complicated. 

To do that, we would need to locate the lombok jar file depending on build system (maven/gradle/etc.). In other editors like IntelliJ or VSCode, this seems to be done by additional configuration or plugin.

How about leaving it as is, and letting users decide what to do with jvm args? That is also the way Emacs LSP has chosen to go: https://emacs-lsp.github.io/lsp-java/#supported-settings

```
lsp-java-vmargs - Specifies extra VM arguments used to launch the Java Language Server. Eg. use -noverify -Xmx1G -XX:+UseG1GC -XX:+UseStringDeduplication to bypass class verification,increase the heap size to 1GB and enable String deduplication with the G1 Garbage collector
```

> What’s noverify do?

`-noverify` makes JVM bypass bytecode verification. `jdt.ls` runs with local files so that's usually not needed. That is my understanding of `-noverify`, but actually I just brought it from the README of `jdt.ls`: https://github.com/eclipse/eclipse.jdt.ls/blob/master/README.md#running-from-the-command-line

**Comment by mjaewon** [comment]
> This seems hacky. Feels like it’s easy to break. Can we have CI dowload the jar to a known location, or something?

Oh maybe you're talking about CI test pipeline? Then how about actually having a lombok jar file under ` ycmd/tests/java/testdata/lombok_project/lib` like below?

having-lombok

**Comment by puremourning** [comment]
> Oh maybe you're talking about CI test pipeline?

Yes I am.

> how about actually having a lombok jar file under

I don't love it. It means that every ycmd user now syncs that file every time. would prefer that the CI script just downloads it to $HOME/lombock as part of the bootstrapping.

**Review by bstaletic — CHANGES_REQUESTED** [review]
@mjaewon Mind using Reviewable for the review? It's much nicer than github reviews and using it consistently correctly groups discussions.

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MSSbHc_35CQyzvOIfDI:b-97ocdo)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic, @mjaewon, and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSSaZM-9VWm-mZfCtJl:b-j3orem) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
>> Any reason for this nested lookup? Is there anything else we could expect to put in 'jvm''s value in the future?

> Not really. I will make it flat like jvm_args if you'd think that is better

I think I do, but I'd like to hear @puremourning's thoughts as well.
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 13 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRj45r0EGZZEorsQV-:-MSSb2gK41uIL2TU2xGo:bblktu8) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L13)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

What’s noverify do?
</blockquote></details>

Skips checking the bytecode of "compiled" files.

https://stackoverflow.com/questions/45936475/why-noverify-added-at-the-end-of-jvm-arguments

**Review by puremourning — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic, @mjaewon, and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSSbkbuANIxoKHwXc91:bkc9u9o) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

>> Any reason for this nested lookup? Is there anything else we could expect to put in 'jvm''s value in the future?

> Not really. I will make it flat like jvm_args if you'd think that is better

I think I do, but I'd like to hear @puremourning's thoughts as well.
</blockquote></details>

according to [this](https://github.com/redhat-developer/vscode-java/wiki/Lombok-support), the vscode extension uses `java.jdt.ls.vmargs`.

Initially I thought "let's use what they use so it's easy to remember", but I think it's harder to remember 'vmargs' than 'jvm_args'.

It also occurs to me that we should probably namespace the jdtl.ls-speicifc options as the top-level dictionary is sort of owned by the LanguageServerCompleter:

```python
{
  # "owned" by the base LanguageServerCompleter...
  "ls": { 
    # existing jdt.ls settings passed on initialise
  },
  "config_sections": { 
    # existing config sections
  },
  "formatting_options": {
    # ...
  },
  # existing top-level stuff
  "project_directory": ...,
  
  # new section for server-specific options that the base LanguageServerComplete won't stomp on
  "server": {
    "jvm_args": [ "...", " ... ", " ... " ],
  }
}
```

I guess when I [wrote this patch](https://github.com/puremourning/ycmd-1/commit/8f849b0a3fb1ab9fffee4e543e78aa89b754420a), I was inclined to use `"jvm": { "args": ... }`. But now I think the nesting is just a bit awkward.

it's quite bikesheddy at this point, but @bstaletic wdyt ?
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MSSfuRa4SWKZmHL6aze:b-29gx67) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

This seems hacky. Feels like it’s easy to break. Can we have CI dowload the jar to a known location, or something?
</blockquote></details>

I was thinking the CI would need to do this, but It occurs to me that anyone running the tests needs it, so how about the run_tests.py downloads it to third-party/lombok similar to how we download libclang and stuff. or maybe we make run_tests.py require gradle to be run in this directory and stick with your approach here?

I don't love requiring gradle any more than we need to. it's a total nuisance in CI, particularly on windows. so if we can make the test runner download it to a known location, then I think it works well enough.

**Review by bstaletic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic, @mjaewon, and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSSkmItAtWQ-oPM1_9B:b71y921) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

according to [this](https://github.com/redhat-developer/vscode-java/wiki/Lombok-support), the vscode extension uses `java.jdt.ls.vmargs`.

Initially I thought "let's use what they use so it's easy to remember", but I think it's harder to remember 'vmargs' than 'jvm_args'.

It also occurs to me that we should probably namespace the jdtl.ls-speicifc options as the top-level dictionary is sort of owned by the LanguageServerCompleter:

```python
{
  # "owned" by the base LanguageServerCompleter...
  "ls": { 
    # existing jdt.ls settings passed on initialise
  },
  "config_sections": { 
    # existing config sections
  },
  "formatting_options": {
    # ...
  },
  # existing top-level stuff
  "project_directory": ...,
  
  # new section for server-specific options that the base LanguageServerComplete won't stomp on
  "server": {
    "jvm_args": [ "...", " ... ", " ... " ],
  }
}
```

I guess when I [wrote this patch](https://github.com/puremourning/ycmd-1/commit/8f849b0a3fb1ab9fffee4e543e78aa89b754420a), I was inclined to use `"jvm": { "args": ... }`. But now I think the nesting is just a bit awkward.

it's quite bikesheddy at this point, but @bstaletic wdyt ?
</blockquote></details>

I don't have a particularly strong opinion. I'm fine with `'server': { 'jvm_args': [] }`. I don't really like `java.jdt.ls.vmargs.omg.why.so.many.damn.stupid.dots.when.will.this.key.name.end`.

**Review by mjaewon — COMMENTED** [review]
@bstaletic Sure thing, just joined Reviewable. Seems I've already messed it up, sorry about that, and I will stick with Reviewable from now one.

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MSUNxUzD2w1CmwqABNB:b-wwlrn0)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSUOqPcAFrCRNDISId5:bksjf4i) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

I don't have a particularly strong opinion. I'm fine with `'server': { 'jvm_args': [] }`. I don't really like `java.jdt.ls.vmargs.omg.why.so.many.damn.stupid.dots.when.will.this.key.name.end`.
</blockquote></details>

Would this work?

```python
return ( settings.get( 'server' ) or {} ).get( 'jvm_args' ) or []
```
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MSUQ7HuCem1RBYIo7v3:bey096u) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I was thinking the CI would need to do this, but It occurs to me that anyone running the tests needs it, so how about the run_tests.py downloads it to third-party/lombok similar to how we download libclang and stuff. or maybe we make run_tests.py require gradle to be run in this directory and stick with your approach here?

I don't love requiring gradle any more than we need to. it's a total nuisance in CI, particularly on windows. so if we can make the test runner download it to a known location, then I think it works well enough.
</blockquote></details>

Thanks. I will make `run_tests.py` download lombok.jar to a specific location then.
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 13 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRj45r0EGZZEorsQV-:-MSUSTu_0UzEJYMpZBDe:b-67obzr) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L13)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Skips checking the bytecode of "compiled" files.

https://stackoverflow.com/questions/45936475/why-noverify-added-at-the-end-of-jvm-arguments
</blockquote></details>

In addition, I brought it from jdt.ls readme.

https://github.com/eclipse/eclipse.jdt.ls/blob/master/README.md#running-from-the-command-line

**Review by bstaletic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic, @mjaewon, and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSUUW9M82bCYmBPodrM:bffy7rv) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
<details><summary><i>Previously, mjaewon (Myungjae Won) wrote…</i></summary><blockquote>

Would this work?

```python
return ( settings.get( 'server' ) or {} ).get( 'jvm_args' ) or []
```
</blockquote></details>

Why not, but you can simplify it a bit.

    return settings.get( 'server', {} ).get( 'jvm_args', [] )

**Review by mjaewon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[ycmd/completers/java/java_completer.py, line 436 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRg0lt5SDljQKGOkTq:-MSuO1Jw8f9sdiO0iuYf:b-ofh2r6) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/completers/java/java_completer.py#L436)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Why not, but you can simplify it a bit.

    return settings.get( 'server', {} ).get( 'jvm_args', [] )
</blockquote></details>

Thanks, fixed as suggested in 5bf05f4
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MSuO6d60huVl3c2gd8_:bxdekof) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
<details><summary><i>Previously, mjaewon (Myungjae Won) wrote…</i></summary><blockquote>

Thanks. I will make `run_tests.py` download lombok.jar to a specific location then.
</blockquote></details>

I wanted to follow the convention (if any) to download a jombok jar file. And it seems libclang is currently being downloaded by this CMakeLists https://github.com/ycm-core/ycmd/blob/master/cpp/ycm/CMakeLists.txt But it doesn't feel right to modify this file to download it.

I'm a bit confused here - Could you shed some light on which existing code I can follow to download third-party libraries for test purpose?

**Review by bstaletic — COMMENTED** [review]
> Reviewed 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MTWMGBP1xvshXsZG7jp:biruv5o) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
<details><summary><i>Previously, mjaewon (Myungjae Won) wrote…</i></summary><blockquote>

I wanted to follow the convention (if any) to download a jombok jar file. And it seems libclang is currently being downloaded by this CMakeLists https://github.com/ycm-core/ycmd/blob/master/cpp/ycm/CMakeLists.txt But it doesn't feel right to modify this file to download it.

I'm a bit confused here - Could you shed some light on which existing code I can follow to download third-party libraries for test purpose?
</blockquote></details>

Sorry for the delay.

There isn't really "a convention" for this type of thing. As you have seen, libclang is downloaded through `CMakeLists.txt`, but clangd is downloaded (and a bunch of others) gets downloaded in `build.py`. Finally, for things that are only useful for tests, neither is appropriate and `run_tests.py` should be used. The `run_tests.py` also downloads another thing in the setup of tests.

**Review by mjaewon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @puremourning)

___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 4 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRipff3zgWtlTbkRqs:-MTYWDDJ7mQGUf9W5Ydy:bovbw5a) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L4)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Sorry for the delay.

There isn't really "a convention" for this type of thing. As you have seen, libclang is downloaded through `CMakeLists.txt`, but clangd is downloaded (and a bunch of others) gets downloaded in `build.py`. Finally, for things that are only useful for tests, neither is appropriate and `run_tests.py` should be used. The `run_tests.py` also downloads another thing in the setup of tests.
</blockquote></details>

Thanks. Just made `run_tests.py` download lombok in this commit a99d064.

**Review by bstaletic — CHANGES_REQUESTED** [review]
> Reviewed 5 of 9 files at r1, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @mjaewon and @puremourning)

___
*[run_tests.py, line 347 at r4](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MT_2YWtECsqJcGAVhgC:-MT_2YWtECsqJcGAVhgD:bcrilw) ([raw file](https://github.com/ycm-core/ycmd/blob/a99d0645fea98159946e6f4ba023d676ca6b1381/run_tests.py#L347)):*
> ```Python
>   LOMBOK_JAR_URL_FORMAT = ( 'https://projectlombok.org/downloads/'
>                             '{lombok_jar_name}' )
>   LOMBOK_JAR_NAME_FORMAT = ( 'lombok-{lombok_version}.jar' )
> ```

This one doesn't need parenthesis since it fits on a single line.

**Review by mjaewon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (waiting on @bstaletic and @puremourning)

___
*[run_tests.py, line 347 at r4](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MT_2YWtECsqJcGAVhgC:-MTbMNCR7PJkrNOZ00mU:b-80ui6l) ([raw file](https://github.com/ycm-core/ycmd/blob/a99d0645fea98159946e6f4ba023d676ca6b1381/run_tests.py#L347)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

This one doesn't need parenthesis since it fits on a single line.
</blockquote></details>

Thank you for the review. Fixed in 8c01dec

**Review by bstaletic — APPROVED** [review]
Generally :lgtm:, but would you mind squashing the commits?

> Reviewed 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MTvVY0x3ZBYxZNOfcmJ:bexr0bm)* status: 1 of 2 LGTMs obtained (waiting on @puremourning)

**Review by mjaewon — COMMENTED** [review]
Not a problem. Just did it.

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MTyDXM9-tmXYC1XYQ01:b-fyjl5e)* status: 0 of 2 LGTMs obtained (and 1 stale) (waiting on @puremourning)

**Review by puremourning — CHANGES_REQUESTED** [review]
Very nice. Very nice indeed!

Just one thing - I think we can avoid another round trip to the user's extra conf here... at least it looks that way to me.

> Reviewed 5 of 9 files at r1, 1 of 2 files at r3, 2 of 3 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MU05dXm92ky_Ykc8cYU:b303819)* status: 0 of 2 LGTMs obtained (and 1 stale) (waiting on @mjaewon)

___
*[run_tests.py, line 342 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU01zWu-waKmY8FyEW9:-MU01zWu-waKmY8FyEWA:bjamltn) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/run_tests.py#L342)):*
> ```Python
> 
> def SetUpJavaCompleter():
>   LOMBOR_DIR = p.join( DIR_OF_THIRD_PARTY, 'lombok', )
> ```

need to add this to the `.gitignore`
___
*[run_tests.py, line 349 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU028Vg2VewfOhBhQf8:-MU028Vg2VewfOhBhQf9:b-43apjc) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/run_tests.py#L349)):*
> ```Python
>   LOMBOK_JAR_NAME_FORMAT = 'lombok-{lombok_version}.jar'
> 
>   jar_name = LOMBOK_JAR_NAME_FORMAT.format( lombok_version = LOMBOK_VERSION )
> ```

prefer to use `f-string`s here, as in:

`jar_name = f'lombok-{ lombok-version }.jar`
`url = f'https://project......../downloads/[ jar_name'`

also note spaces inside `{ curly expressions }`
___
*[ycmd/completers/java/java_completer.py, line 432 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU04Pz87cwP2rgcNqgB:-MU04Pz87cwP2rgcNqgC:bltpvsm) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/ycmd/completers/java/java_completer.py#L432)):*
> ```Python
> 
> 
>   def _GetJvmArgs( self, request_data ):
> ```

By the time this is called, the base class `LanguageServerCompleter` has already called `self._GetSettingsFromExtraConf` (as part of `_StartAndInitializeServer`.

TL;DR, I *think* that means we already have the return from `self.GetSettings` in `self._settings` at this point, so we can avoid calling the user's Settings function again, and just use `self._settings.get( 'server', {} ).get( 'jvm_args', [] )` in place when we start the server.
___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 13 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRj45r0EGZZEorsQV-:-MU05VrA71htBZV6ZdKI:bq8o7ly) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L13)):*
<details><summary><i>Previously, mjaewon (Myungjae Won) wrote…</i></summary><blockquote>

In addition, I brought it from jdt.ls readme.

https://github.com/eclipse/eclipse.jdt.ls/blob/master/README.md#running-from-the-command-line
</blockquote></details>

That's a pretty good source!

**Review by mjaewon — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529)* status: 0 of 2 LGTMs obtained (and 1 stale) (waiting on @bstaletic and @puremourning)

___
*[run_tests.py, line 342 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU01zWu-waKmY8FyEW9:-MU0MIPu6oLcTMXsv64d:bn0lvpw) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/run_tests.py#L342)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

need to add this to the `.gitignore`
</blockquote></details>

Thanks for pointing that out. Fixed in 03db1cb 
___
*[run_tests.py, line 349 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU028Vg2VewfOhBhQf8:-MU0MNZs1TdrwUPjMYNv:bxc878w) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/run_tests.py#L349)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

prefer to use `f-string`s here, as in:

`jar_name = f'lombok-{ lombok-version }.jar`
`url = f'https://project......../downloads/[ jar_name'`

also note spaces inside `{ curly expressions }`


</blockquote></details>

Fixed in 03db1cb as suggested
___
*[ycmd/completers/java/java_completer.py, line 432 at r6](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MU04Pz87cwP2rgcNqgB:-MU0MRbj2mXJVNKcng1u:b-af23u2) ([raw file](https://github.com/ycm-core/ycmd/blob/1a0f11d5ac76cfa93560b75048b74021e3c5873f/ycmd/completers/java/java_completer.py#L432)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

By the time this is called, the base class `LanguageServerCompleter` has already called `self._GetSettingsFromExtraConf` (as part of `_StartAndInitializeServer`.

TL;DR, I *think* that means we already have the return from `self.GetSettings` in `self._settings` at this point, so we can avoid calling the user's Settings function again, and just use `self._settings.get( 'server', {} ).get( 'jvm_args', [] )` in place when we start the server.
</blockquote></details>

Thanks for the detailed explanation! Didn't know that. Fixed in 03db1cb

**Review by puremourning — APPROVED** [review]
:lgtm:

Thanks!

This has been a very high quality PR and I really appreciate the work and diligence and positive attitude you've shown here. It's been a pleasure to review. Much appreciated.

> Reviewed 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MU0Rxg89vcCJtMZlggH:blkcwvw)* status: 1 of 2 LGTMs obtained (and 1 stale) (waiting on @mjaewon)

**Review by mjaewon — COMMENTED** [review]
Thank you for saying like that. Glad that I helped.

What would be the next step?

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MU0XnmN9uKR93tf_3Hg:b-24m5wr)* status: 1 of 2 LGTMs obtained (and 1 stale) (waiting on @mjaewon)

**Comment by puremourning** [comment]
Just need @bstaletic  to sign off on it and the bot will merge it.

**Review by bstaletic — APPROVED** [review]
Thanks once again. Let's ship this. :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/ycm-core/ycmd/1529#-:-MU0l61P3t2bDmDvOB6M:b-umqh42)* status: :shipit: complete! 2 of 2 LGTMs obtained (waiting on @mjaewon)

___
*[ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py, line 13 at r1](https://reviewable.io/reviews/ycm-core/ycmd/1529#-MSRj45r0EGZZEorsQV-:-MU0lSgPC7BcDijRr2Pt:bmbq78u) ([raw file](https://github.com/ycm-core/ycmd/blob/4cf10f5e35317c543abddb3e1d7d4b9d6dd5c648/ycmd/tests/java/testdata/lombok_project/.ycm_extra_conf.py#L13)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

That's a pretty good source!
</blockquote></details>

So we're saying we trust eclipse? :D

**Comment by bstaletic** [comment]
Come on, you stupid bot! :lgtm_strong:
